server/datastore/mysqlredis/hosts.go

```diff
@@ -0,0 +1,140 @@
+package mysqlredis
+
+import (
+	"context"
+	"time"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/contexts/logging"
+	"github.com/fleetdm/fleet/v4/server/datastore/redis"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	redigo "github.com/gomodule/redigo/redis"
+)
+
+const enrolledHostsSetKey = "enrolled_hosts:host_ids"
+
+// TODO: additionally, we could have a cron job to re-sync the counts every so
+// often, it would:
+// a) SELECT COUNT(*) the hosts in the DB and compare with SCARD in redis
+// b) only if different (could even be only if "significantly" different), load
+// the IDs and replace the redis set with those.
+//
+// For additional safety, we could check the DB if we're about to reject an
+// enrollment: if so, check the COUNT(*) in the DB, and if it matches or is
+// above, reject, if it doesn't match load the IDs and replace the redis SET
+// with those.
+
+func (d *datastore) addHosts(ctx context.Context, hostIDs ...uint) error {
+	conn := redis.ConfigureDoer(d.pool, d.pool.Get())
+	defer conn.Close()
+
+	args := redigo.Args{enrolledHostsSetKey}
+	args = args.AddFlat(hostIDs)
+	_, err := conn.Do("SADD", args...)
+	return ctxerr.Wrap(ctx, err, "enrolled limits: add hosts")
+}
+
+func (d *datastore) removeHosts(ctx context.Context, hostIDs ...uint) error {
+	conn := redis.ConfigureDoer(d.pool, d.pool.Get())
+	defer conn.Close()
+
+	args := redigo.Args{enrolledHostsSetKey}
+	args = args.AddFlat(hostIDs)
+	_, err := conn.Do("SREM", args...)
+	return ctxerr.Wrap(ctx, err, "enrolled limits: remove hosts")
+}
+
+func (d *datastore) checkCanAddHost(ctx context.Context) (bool, error) {
+	conn := redis.ConfigureDoer(d.pool, d.pool.Get())
+	defer conn.Close()
+
+	n, err := redigo.Int(conn.Do("SCARD", enrolledHostsSetKey))
+	if err != nil {
+		return false, ctxerr.Wrap(ctx, err, "enrolled limits: check can add host")
+	}
+	if n >= d.enforceHostLimit {
+		// TODO(mna): check in DB to make absolutely sure the number is correct?
+		return false, nil
+	}
+	return true, nil
+}
+
+func (d *datastore) NewHost(ctx context.Context, host *fleet.Host) (*fleet.Host, error) {
+	if d.enforceHostLimit > 0 {
+		ok, err := d.checkCanAddHost(ctx)
+		if !ok && err == nil {
+			return nil, ctxerr.Errorf(ctx, "maximum number of hosts reached: %d", d.enforceHostLimit)
+		}
+		if err != nil {
+			logging.WithErr(ctx, err)
+		}
+	}
+
+	h, err := d.Datastore.NewHost(ctx, host)
+	if err == nil && d.enforceHostLimit > 0 {
+		if err := d.addHosts(ctx, h.ID); err != nil {
+			logging.WithErr(ctx, err)
+		}
+	}
+	return h, err
+}
+
+func (d *datastore) EnrollHost(ctx context.Context, osqueryHostID, nodeKey string, teamID *uint, cooldown time.Duration) (*fleet.Host, error) {
```

Comment:
I'll see how that would look. I agree about the principle but I'm afraid this might end up being significantly more complex, since the service needs to be aware of whether the enforcement is enabled or not (instead of being abstracted away in the Datastore from the moment `serve` is called), and it needs a Datastore method to call to get the count - which (awkwardly) would be implemented only by the redis wrapper datastore, being a no-op elsewhere - just so it can test the condition and fail if it's over the limit.

(arguably the non-redis-wrapper datastore could return `0, nil` so that the service does not need to know if enforcement is enabled, but then the datastore method is plain wrong if enforcement is not enabled, e.g. calling `Datastore.GetEnrolledHostsCount()` returns a number that may or may not be right depending on whether the check is enabled)

Reply:
> the service needs to be aware of whether the enforcement is enabled or not (instead of being abstracted away in the Datastore from the moment fleet serve is called)

The `Service` already has access to `FleetConfig`, so this would be an if check. We would need to pass the `deviceCount` though.

> needs a Datastore method to call to get the count - which (awkwardly) would be implemented only by the redis wrapper datastore, being a no-op elsewhere

We don't have to use the wrapping method here. We could just pass a redis connection to the datastore in `mysql.New(...)`. `GetEnrolledHostsCount` would only be called by the service if it needs to. Same goes for incrementing the enrolled hosts count. The service would be responsible for calling these datastore methods appropriately.

Reply:
> The Service already has access to FleetConfig, so this would be an if check. We would need to pass the deviceCount though.

True, it also has the license info so it does have the `deviceCount`.

> We could just pass a redis connection to the datastore in mysql.New(...).

This I'm not convinced though. I'd still use the wrapping datastore so the redis access is only done in the redis wrapper.

> Same goes for incrementing the enrolled hosts count. The service would be responsible for calling these datastore methods appropriately.

I disagree that this is cleaner than the wrapped Datastore, I think it's much nicer to have the wrapper take care transparently of keeping track of the counts rather than add this in all related Service methods (plus the cleanup calls are not even triggered by a service method, it's in a cron job). It's not only nicer, it's safer - if the datastore methods end up being called elsewhere, they will still maintain proper counts. I agree that the actual count validation check makes sense in the `Service`, but not maintaining the redis set IMHO.